Vault: Revamp audit logging docs (#631)
hashicorp/web-unified-docs: content/vault/v1.20.x/content/docs/audit/index.mdx

description: Audit devices are mountable devices that log requests and responses in Vault.
---

# Audit logging

> Audit devices enable the detailed recording of all Vault API requests and responses, except for a small set of [API endpoints](#excepted-api-endpoints). An audit device can be a [file](/vault/docs/audit/file), a [syslog server](/vault/docs/audit/syslog), or a [socket](/vault/docs/audit/socket).

Reviewer: ```suggestion
With a a small set of [exceptions](#excepted-api-endpoints), Vault audit devices enable the detailed recording of all API requests and responses. Audit devices can be [files](/vault/docs/audit/file), a [syslog server](/vault/docs/audit/syslog), or a [socket](/vault/docs/audit/socket).
```
Better to highlight the exceptions first, instead of claiming you can audit "everything" and then contradicting ourselves.

> Refer to the [best practices](/vault/docs/audit/best-practices) page for important elements to consider when configuring Vault's audit devices.

Reviewer: ```suggestion
Refer to the [best practices](/vault/docs/audit/best-practices) page for important considerations before configuring your audit devices.
```

> <Tip>

Reviewer: ```suggestion
```
This isn't a tip, it's just more information about audit devices

>   Audit logs are distinct from server logs (also known as operational logs). Server logs emit information about the Vault server process, such as process errors and configuration warnings. Audit logs record the details of every request and response received and sent by the Vault API. Refer to the <a href="/vault/tutorials/monitoring/troubleshooting-vault#vault-server-logs">Troubleshoot Vault</a> guide for more information about server logs.

Reviewer: ```suggestion
Audit logs are distinct from server logs (also known as operational logs).
Server logs emit information about the Vault server process, such as process
errors and configuration warnings. Audit logs record the details of every
request received and response and sent by the Vault API.

Refer to the
[Troubleshoot Vault](/vault/tutorials/monitoring/troubleshooting-vault#vault-server-logs)
tutorial for more information about server logs.
```

> </Tip>

Reviewer: ```suggestion
```

## Enabling and disabling audit devices

> When a new Vault cluster is initialized, no auditing is enabled. Audit devices must be enabled via the [API](/vault/api-docs/system/audit#enable-audit-device) or related tooling (such as the Vault CLI or the Terraform Vault provider).

Reviewer: ```suggestion
When a new Vault cluster initializes, auditing is disabled. You must explicitly
enable audit devices with the
[Vault CLI](/vault/docs/commands/audit/enable),
[API](/vault/api-docs/system/audit#enable-audit-device), or the
[Terraform Vault provider](/vault/docs/configuration/programmatic-management).
```
Style correction: write in active voice

> When you initialize a new Vault cluster, auditing is disabled. You must enable audit devices with the [API](/vault/api-docs/system/audit#enable-audit-device) or related tooling (such as the [Vault CLI](/vault/docs/commands/audit/enable) or the [Terraform Vault provider](/vault/docs/configuration/programmatic-management)).

Author: Fixed in 264bcf1046282a780c7c027c786c646378b762bf

> When enabling an audit device, options can be passed to it to configure it. There is a set of [common configuration options](/vault/api-docs/system/audit#common-configuration-options) that apply to all types of audit devices. In addition, each type of audit device has its own set of options; refer to the documentation on the left for each type.

Reviewer: ```suggestion
Each audit device has unique configuration options in addition to the
[common configuration options](/vault/api-docs/system/audit#common-configuration-options).
```
Style correction: write simply

> For example, the command below enables the file audit device and configures it to write to the file at the `file_path` location, to disable [hashing](#hashing-of-sensitive-values) of token accessors, and to [elide list responses](#eliding-list-response-bodies):

Reviewer: ```suggestion
For example, the command below enables a `file` type audit device that writes to
the file at `/var/log/vault_audit.log`. The additional options disable
[hashing](#hashing-of-sensitive-values) of token accessors, and enables
[elide list responses](#eliding-list-response-bodies).
```
Style correction: write in active voice

$ vault audit enable file \
    file_path=/var/log/vault_audit.log \
    hmac_accessor=false \
    elide_list_responses=true
```

> Vault replicates audit device configurations to all nodes within a cluster. In Vault Enterprise, Vault also replicates audit device configurations to Performance Secondary and DR Secondary clusters unless you configure the audit device as [`local`](/vault/api-docs/system/audit#local) to the cluster.

Author: Fixed in 264bcf1046282a780c7c027c786c646378b762bf

<Note>
  Before you enable an audit device, ensure that all Vault nodes to which the audit device configuration replicates can successfully log to that audit device.

> </Note>

Reviewer: It doesn't really belong on this page because the page is about best practices and the note isn't a recommendation, it's actual instruction. That's why I suggested adding it to the CLI and API pages for enabling audit devices instead. Unless they have the command/endpoint fully memorized, they'll likely look at those references pages and should still see the information before enabling.

> </Note>

Reviewer: To be frank, it *should* be in a how-to guide. We just don't have one at the moment. I'm okay leaving it here for now, but it should be worked into the rest of the content not fenced off as an aside.

When you disable an audit device, it stops receiving logs immediately. The existing logs that it stored remain untouched.

<Warning>
  Once you disable an audit device, you can no longer <a href="#hashing-of-sensitive-values">compute hashes</a> for audit log entries previously written to that audit device. This remains true even if you re-enable the audit device at the same path, as Vault creates a new hashing key.
</Warning>

## Availability of audit devices  

Audit logs are critically important, and ignoring auditing failures opens an avenue for attack. As such, when you enable at least one audit device, Vault guarantees that it records every API request and response to *at least one* of the enabled audit devices. If at least one audit device is enabled *and* Vault cannot write to at least one of the enabled audit devices, Vault refuses to service the corresponding API request.

<Note>

>   When all enabled audit devices become unavailable, Vault in effect becomes unavailable as well. To prevent audit-related unavailability of Vault, HashiCorp recommends that you enable at least two audit devices. Refer to the [best practices](/vault/docs/audit/best-practices) page for additional information.

Author: Fixed in 264bcf1046282a780c7c027c786c646378b762bf. Noting that I left out the second recommendation from this suggestion, as it's unrelated to the availability of audit devices.

</Note>

When two or more audit devices are enabled, Vault attempts to send each audit log entry to all audit devices. However, Vault guarantees only that it writes an audit log entry to *at least one* audit device. Hence, to ensure a complete record of every API request and response when multiple audit devices are enabled, you must analyze the combined audit log entries across all enabled audit devices.

## Hashing of sensitive values ((#hashing-of-sensitive-values))

> To protect the confidentiality of potentially sensitive information in audit logs, Vault by default writes only a keyed hash (HMAC-SHA256) of most string values within request and response audit entries. Vault exposes an API endpoint, [`/sys/audit-hash`](/vault/api-docs/system/audit-hash), to calculate the keyed hash of any known value for a given audit device.

Author: Took a slightly different approach but addressed in 264bcf1046282a780c7c027c786c646378b762bf

The [`elide_list_responses`](/vault/api-docs/system/audit#elide_list_responses) option replaces `keys` and `key_info` lists in audit log entries for responses to list requests with the number of objects they contain. This improves compatibility with audit devices and reduces the amount of audit data that is ingested, stored, and processed, while retaining some level of information about what was returned to the client.

Here is an example of an audit record with elided lists (formatted with extra whitespace, and with fields not relevant to the example omitted):

```json
{
  "type": "response",
  "request": {
    "operation": "list"
  },
  "response": {
    "data": {
      "key_info": 4,
      "keys": 4 // instead of ["a", "b", "c", "d"]
    }
  }
}
```

## Logging of request headers

> You can use the [`/sys/config/auditing`](/vault/api-docs/system/config-auditing#create-update-audit-request-header) API to configure which headers in incoming client requests Vault records in the corresponding audit log entries.

Reviewer: ```suggestion
You can use the
[`/sys/config/auditing`](/vault/api-docs/system/config-auditing#create-update-audit-request-header)
endpoint to configure which incoming client request headers you want Vault to
record in the corresponding audit log entries.
```

> By default, Vault automatically records values for the User Agent header, as well as for the `correlation-id` and `x-correlation-id` request headers to assist with request correlation in distributed systems.

Reviewer: ```suggestion
By default, Vault records the following request headers to assist with request
correlation in distributed systems:

- `User-Agent`
- `Correlation-Id`
- `X-Correlation-Id`
```
Style correction: make important information easy to scan, write in active voice

> Vault does not hash request header values by default, including for the <code>correlation-id</code> and <code>x-correlation-id</code> headers. You can [configure](/vault/api-docs/system/config-auditing#hmac) Vault to apply the [keyed hash](#hashing-of-sensitive-values) to the values of request headers you consider sensitive.

Reviewer: ```suggestion
Vault does not hash request header values by default. You can
[configure](/vault/api-docs/system/config-auditing#hmac) Vault to apply the
[keyed hash](#hashing-of-sensitive-values) to any request header value you
consider sensitive.
```
Style correction: simplify sentences when possible

> ## Excepted API endpoints ((#excepted-api-endpoints))

Reviewer: ```suggestion
## Exempted API endpoints ((#exempted-api-endpoints))
```
For accessibility purposes, the better word here is "exempted" instead of "excepted" which is a close homophone of "accepted" and might create confusion for folks using screen readers . We also need to update the places that link to the old anchor

> Vault does not record requests and responses to and from the following API endpoints to its audit devices:

Reviewer: ```suggestion
Vault does not audit requests and responses for the following API endpoints:
```
Style correction: simplify sentences whenever possible

```
sys/init
sys/seal-status
sys/seal
sys/step-down
sys/unseal
sys/leader
sys/health
sys/rekey/init
sys/rekey/update
sys/rekey/verify
sys/rekey-recovery-key/init
sys/rekey-recovery-key/update
sys/rekey-recovery-key/verify
sys/storage/raft/bootstrap
sys/storage/raft/join
sys/internal/ui/feature-flags
```

> Additionally, Vault does not record requests to the following API endpoints if unauthenticated access to them is permitted by the relevant listener configuration settings:

Reviewer: ```suggestion
Additionally, Vault does not audit requests to the following endpoints when the
associated listener configuration allows unauthenticated access:
```

```
sys/metrics
sys/pprof/*
sys/in-flight-req
```

> ## Tutorial

Author: Never mind: just saw the suggested changes to the filtering docs. Will address in an upcoming commit.
